examples/dcutr/src/main.rs

```diff
@@ -172,7 +172,7 @@ fn main() -> Result<(), Box<dyn Error>> {
     .build();
 
     swarm
-        .listen_on("/ip4/0.0.0.0/udp/0/quic-v1".parse().unwrap())
+        .listen_on("/ip4/127.0.0.1/udp/0/quic-v1".parse().unwrap())
```

Comment:
You said this is only because of how we implemented `eligible_listener` right?

But if you only listen on localhost, then this example is sort of use-less. Hole-punching on localhost doesn't make much sense and if we don't listen on all interfaces, you can't actually use this example to test hole-punching.

Reply:
In such case the `eligible_listener` implementation needs to be changed, no?

Reply:
Frankly speaking I don't know what purpose serves that `listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()`. It was introduced in https://github.com/libp2p/rust-libp2p/pull/2289/commits/1a4b8acc80028ae2ac221224caf4ad9b3ceb95c1#diff-b1dcda9367774dd8a742457ff79bc528d030a8cc38d6791efefc78b30f3f2e03R141

Reply:
My guess is that it forces that the local addr of an "outgoing" packet with destination localhost uses a localhost listening socket, if present.

I believe @elenaf9 originally implemented this, maybe she can shed some light on this :)

Reply:
IIRC I copied that from the TCP logic for selecting a local listener suitable for dialing: https://github.com/libp2p/rust-libp2p/blame/73082212374a420d5cb095df6fcf2888249f1dbc/transports/tcp/src/lib.rs#L139

Reply:
@kpp I don't understand why the `listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()` is a problem. Can you please expand?

Reply:
Friendly ping. With my testing via the Rust punchr client, I think this can be reverted. @kpp any objections?

Reply:
That's weird we can hole punch `!is_loopback()` remote addrs with `0.0.0.0` but we can't punch `127.0.0.1` addrs. Maybe we should remove this `listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()` check?

Reply:
The issue is simple, we have have the following setup:

```
cargo r -p relay-server-example -- --secret-key-seed 1 --port 15353

cargo r -p dcutr -- --mode listen --secret-key-seed 2 --relay-address /ip4/127.0.0.1/udp/15353/quic-v1/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X

cargo r -p dcutr -- --mode dial --secret-key-seed 3 --relay-address /ip4/127.0.0.1/udp/15353/quic-v1/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X --remote-peer-id 12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3
```

And it does not work without my patch, it fails with "no listeners" error. First of all, this code does not work on master and I am pretty surprised. Do you thing the setup above is wrong? If yes, lets fix my setup. If it is a correct setup then how shall we make the code work?

Reply:
When I run the commands above the third node successfully "hole punches" to the second node.

```
[2023-07-27T13:06:46Z INFO  dcutr] DirectConnectionUpgradeSucceeded { remote_peer_id: PeerId("12D3KooWH3uVF6wv47WnArKHk5p6cvgCJEb74UTmxztmQDc298L3") }         
```

Though note, I am saying "hole punches" in quotes as there is nothing to hole punch to. In others words, there is no firewall or NAT given that all 3 nodes are running on my local machine.

I am not sure why the above does not work on your machine. I.e. I am not sure why you get a "no listeners" error.

Reply:
Thanks for being persistent here @kpp. This indeed seems to be a bug.

In TCP we track the specific listen addresses, even if the user calls `listen_on` with an unspecified address (e.g. `0.0.0.0`). If the user calls `listen_on` with a specific IP address i.e. not a wildcard like `0.0.0.0` but e.g. `127.0.0.1` we track (here _register_) the specific address right away.

https://github.com/libp2p/rust-libp2p/blob/f10f1a274abe6e5193de653aa4dd951ae56c775e/transports/tcp/src/lib.rs#L394

In case where the user calls `listen_on` with an unspecified address (e.g. `0.0.0.0`) we depend on `if-watch`, more specifically `poll_if_watch` to provide us with the concrete addresses:

https://github.com/libp2p/rust-libp2p/blob/f10f1a274abe6e5193de653aa4dd951ae56c775e/transports/tcp/src/lib.rs#L666-L682 

Either way, we end up tracking specific listen addresses with port-reuse enabled and can thus choose the right listener in `local_dial_addr`:

https://github.com/libp2p/rust-libp2p/blob/f10f1a274abe6e5193de653aa4dd951ae56c775e/transports/tcp/src/lib.rs#L122-L151

In `libp2p-quic` we only track the initial address that the user provided on the `listen_on` call.

https://github.com/libp2p/rust-libp2p/blob/f10f1a274abe6e5193de653aa4dd951ae56c775e/transports/quic/src/endpoint.rs#L170-L178

This might be a wildcard address. Later on, when dialing a `127.0.0.1` address, the check `listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()` discards the listener with the unspecified listen address, thus a new dial-only socket is used. Using a new socket for dialing makes hole punching from the dcutr dialer to the dcutr listener impossible.

How to move forward?

I am not sure simply removing the `listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()` check is the way to go. This check is still helpful. E.g. say we have two listeners, one on localhost, one not on localhost. Ideally we would use the former when dialing.

We could mirror what `libp2p-tcp` does. That is, track the specified listen addresses reported by if-watch. Later, when dialing, choose the listener with the right address.

Ideally I would like to have the operating system choose. It has the most information about all available routes to each interface.

For now, I suggest we go with the `libp2p-tcp` approach.

Reply:
Thinking about it some more, this is not a bug introduced in this pull request, but instead an already existing bug in our `libp2p-quic` implementation. With that in mind, I don't think this pull request needs to block on a fix for it.

I documented it in https://github.com/libp2p/rust-libp2p/issues/4259 instead. Let's revert the `dcutr/main.rs` change in this pull request and tackle the root issue in a follow-up pull request.